# Курс TextAI

# Материалы: https://bit.ly/2023-TextAI

## Введение в Python

* Google Colab
* Google Disk
* Python
* Numpy
* Pandas

## Самостоятельная работа

* База: https://app.datacamp.com/learn/courses/intro-to-python-for-data-science
* Pandas: https://app.datacamp.com/learn/courses/data-manipulation-with-pandas
* Функции: https://app.datacamp.com/learn/courses/python-data-science-toolbox-part-1
* Итераторы: https://app.datacamp.com/learn/courses/python-data-science-toolbox-part-2

## Подключаем Google Disk

In [54]:
# МЕНЯ ВСТАВИТЬ В САМОЕ НАЧАЛО ЛЮБОГО JupyterNotebook,
#   А ПРИ ПОВТОРНОЕ РАБОТЕ ВСЕГДА ЗАПУСКАТЬ!

# Тут имя папки, в которую все сохраняется. 
#  Менять это имя можно. 
#  На Google Drive должно быть та же папка
#  Регистр (ЗАГЛАВНЫЕ или строчные буквы) важен! 
FOLDER = "2023-TextAI" 

PATH = "drive/MyDrive/" + FOLDER  + "/"

from google.colab import drive
drive.mount('/content/drive')

import os
os.makedirs(PATH, exist_ok = True)

import sys
sys.path.append(PATH)

Mounted at /content/drive



## Знакомство с Jupyter Notebook

### Интерфейс

Основная часть **Jupyter Notebook** состоит из _блоков_. 

Есть два основных вида блока:
* Текст (Markdown)
* Код (Python в нашем случае)

Время от времени ноутбук автоматически сохраняется. Также это можно сделать вручную **CTRL+S** или с помощью соответствующей пиктограммы на Toolbar'е.

Один (или несколько) блоков всегда выделены.

Ниже такого выделения можно добавить ещё один блок с помощью пиктограммы '+'. Новый блок будет типа "код". Тип блока можно поменять с помощью отдельного контрола на Toolbar'е.

Выделенный блок можно вырезать (блок удаляется, а его содержимое сохраняется) с помощью пиктограммы "ножницы". Также можно просто скопировать содержимое ячейки. Далее содержимое ячейки можно вставить в код ниже выделенного блока, с помощью соответсвующей пиктограммы.

Отдельной кнопки на удаление блока нет. Необходимо его вырезать.

Каждый блок может быть _запущен_: кнопка "Run" или **CTRL+Enter**. Текст превращается в красиво-отформатированный, а код запускается. **Порядок запуска блоков кода важен(!)**, а порядок записи блоков в **Notebook** нет влияет только на визуальный аспект.

В ноутбуке Kernel --- это тот "движек", который понимает код Python и выполняет его. Его можно остановить, или перезапустить, тогда забудутся все вычисления, выполненые до этого, и нужно будет их выполнить заново. Также можно перезапустить Kernel и выполнить весь код в ноутбуке.

Также с помощью меню _File->Download As_ можно сконвертировать файл в HTML, и другие форматы, чтобы показать результат работы людем без Python.

### Markdown

Двойной клик на текстовый блок превращает его в **Markdown**. В частности, в этом блоке можно увидеть как оформляются
1. Заголовки (чем больше "диезов", тем больше уровень вложенности заголовка)
1. Картинки
1. Нумерованный и ненумерованный списки
1. Курсивный и полужирный текст (две звездочки или два подчеркивания -- полжирный, одна(о) -- курсив)
1. Ссылки
1. Формулы

В **Markdown** абзацы отделяются друг от друга пустой строкой. 

[Ссылка на этот ноутбук.](https://yadi.sk/d/_AaMOKWloehPKQ)

Формулы оформляются с помощью **LaTeX**.

* $A^5$
* $B_2$
* $f(x) = \frac{A_0}{2} + \sum \limits_{n=1}^{\infty} A_n \cos \left( \frac{2 n \pi x}{\nu} - \alpha_n \right)$


## Python

### Basics

Переменные создаются с помощью '='. 

Имя переменной может содержать:
* Буквы, заглавные и строчные
* Подчеркивание
* Цифры
* Не может начинаться с цифры

При определении типа переменной Python использует так называемый [duck typing](https://en.wikipedia.org/wiki/Duck_typing).

> Если животное выглядит как лошадь и ведёт себя как лошадь, значит это лошадь.

При именовании переменных обычно применяется snake_case.

In [3]:
#  Строки:
i_am_a_string = "55"
i_am_also_a_string = '66'

i_am_an_int_number = 55
i_am_a_real_number = 55.0

now_i_am_a_number = int(i_am_a_string)
now_i_am_a_string = str(i_am_a_real_number)

i_am_a_boolean = True
i_am_also_a_boolean = (5>6)
# А это ничто, великое и могучее...
i_am_nothing = None

In [4]:
# Только результат последней строки печатается
i_am_an_int_number
i_am_a_real_number

55.0

In [5]:
# Но print работает всегда
print(i_am_an_int_number)
print(i_am_a_real_number)
"DONE"

55
55.0


'DONE'

In [6]:
# Print -- это функция, может принимать несколько аргументов
print(i_am_an_int_number, i_am_a_real_number)

55 55.0


In [7]:
# Просто выполнили операцию, но никуда не сохранили (распечатали только, потому что результат операции является результатом последней команды в ячейке)
i_am_an_int_number + i_am_a_real_number

110.0

In [8]:
i_have_remembered_the_sum = i_am_an_int_number + i_am_a_real_number

In [9]:
# Можем переопределить значение и даже тип переменной
#  Но перепоределять тип крайне не желательно с "человеческой" точки зрения

i_am_an_int_number = "I am not an int, Ah, ah, ah...!"

Есть много операций с базовыми типами, как правило, есть операция не экзотическая, т.е. если она нужна часто, то она уже реализована. 

[Сводная таблица с операциями.](https://www.tutorialspoint.com/python/python_basic_operators.htm)

Google/Yandex тут очень хорошо помогают.

### Условный оператор if

Каждый блок вызывается построчно сверху вниз. Иногда команда должна вызываться только при соблюдении определённых условий. Так мы можем распечатать, является ли число положительным или отрицательным

In [ ]:
a = 5
if a > 0:
    print("Число 'a' -- положительное")
    print("aaa")
elif a < 0:
    print("Число 'a' -- отрицательное")
else:
    print("Число 'a' равно 0")    
        
# Выражение a>0 возвращает логический тип, 
#   в _if_,_elif_ может быть любое другое выражение, возвращающиее логический тип

if True:
    print("Всегда печатаем это")
else:
    print("Никогда не распечатается")

# Добавим ещё проверку на чётность для положительных чисел,
#   а также покажем, что можно менять значение переменной
res = "Число а не проверено"

if a > 0 and (a % 2) == 0:
    res = "Число 'a' -- четное положительное"
elif a > 0 and a % 2 == 1:
    res = "Число 'a' -- нечетное положительное"
elif a < 0:
    res = "Число 'a' -- отрицательное"
else:
    res = "Число 'a' равно 0"
    
print(res)

# Также можно вкладывать один if в другой
res = "Число а не проверено"

if a > 0:
    if a % 2 == 0:
        res = "Число 'a' -- четное положительное"
    else:
        res = "Число 'a' -- нечетное положительное"
elif a < 0:
    res = "Число 'a' -- отрицательное"
else:
    res = "Число 'a' равно 0"
    
print(res)

#### Задание: условный оператор

Ниже представлены задачи, которые можно реализовать с помощью условного оператора.
Условие задач написано в комментарии, ниже комментария идут начальные значения для переменных.
Эти начальные параметры можно менять и нужно сделать так, чтобы код работал.

Кроме того нужно проверить, что входные значения возможны.

In [ ]:
# Пример
# Задача 0: проверить, является ли год YEAR высокосным.
# В рамках данной задачи считать год высокосным, если он делиться на 4 нацело
YEAR=2023

# Проверяем, что год положительный
if YEAR < 0:
    print("Передан невозможный год")
elif YEAR % 4 == 0:
    print(YEAR, "год --- високосный")
else:
    print(YEAR, "год --- невисокосный")

In [ ]:
# Задача 1: проверить, является ли год YEAR высокосным.
# В рамках данной задачи считать год высокосным, 
#  если он делиться на 400 нацело или делиться на 4 и при этом не делиться на 100
YEAR=2023

# ...

## Функции

Функции -- кусок кода, который вызвается, когда вызывается имя функции.
Функции позволяют переиспользовать кусок кода многократно. Это позволяет:

* Передавать другим людям свой код
* Исправлять ошибки в только в одном месте (внутри функции, а не в каждом месте, где используется код)
* Использовать локальные имена переменных

Примером такой функции является функция **print**: 
* Мы её не писали, т.е. с нами поделились этой функцией
* Если в ней обнаружат ошибки, то нам не придётся менять код
* Когда мы пользуемся функцией **print**, мы не задумываемся какие имена переменных были использованы в её реализации, и, наоборот, разработчики этой функции не задумывались о именах переменных в нашем коде.

Поведение функции может быть изменено посредством параметров. Так, например, функция может возвращать, является ли год высокосным в зависимости от того года, который был передан в качестве параметра.

In [ ]:
def is_leap_year(year):
    # Проверяем, что год положительный
    if year < 0:
        return None # Обработка ошибки
    elif year % 4 == 0:
        return True
    else:
        return False
    
# Либо можно упростить
def is_leap_year2(year):
    # Проверяем, что год положительный
    if year < 0:
        return None # Обработка ошибки
    else:
        return year % 4 == 0

In [ ]:
# Вызываем
print(is_leap_year(2023))
print(is_leap_year2(2023))

### Позиционные и именованные переменные

При определении функции, все переменные имеют имя. Некоторые из них могут иметь значение по умолчанию.

При вызове функции, параметры могут передаваться через их имя, либо по позиции. Сначала идут позиционные параметры, потом именнованные.

In [ ]:
def is_leap_year3(year, simplified=True):
    # Проверяем, что год положительный
    if year < 0:
        return None # Обработка ошибки
    elif simplified:
        return year % 4 == 0
    else:
        return year % 400 == 0 or (year % 4 == 0 and year % 100 != 0)

In [ ]:
print(is_leap_year3(2023)) # OK
print(is_leap_year3(2023, True)) # OK
print(is_leap_year3(2023, simplified=True)) # OK
print(is_leap_year3(simplified=True, year=2023))  # OK
print(is_leap_year3(simplified=True)) # не OK


Если поменять имя параметра, то при вызове функции позиционные параметры не поменяются, а для именованных поменяется имя.

In [ ]:
def is_leap_year4(y, simplified=True):
    # Проверяем, что год положительный
    if y < 0:
        return None # Обработка ошибки
    elif simplified:
        return y % 4 == 0
    else:
        return y % 400 == 0 or (y % 4 == 0 and y % 100 != 0)

In [ ]:
print(is_leap_year4(2023))
print(is_leap_year4(2023, True))
print(is_leap_year4(2023, simplified=True))
print(is_leap_year4(simplified=True, y=2023)) # Тут y вместо year

In [ ]:
# Задача: задан номер дня, месяца и года. Написать функцию, которая возвращает является ли дата возможной (True) или нет (False). 
#   Так, например, 1 января 2023 года -- возможная дата, а 31 июня 2023 -- невозможная.
# Воспользуйтесь функцией по определению высокосности года (кстати, зачем она нам тут?)


In [ ]:
# Вызовите написанную функцию выше два раза, один раз с позиционными параметрами, другой раз с именованными
# 1

# 2


## Циклы: начало

Цикл повторяет определённое действие много раз, что напоминает функцию. Отличием является то, что, с одной стороны, переиспользовать код внутри цикла нельзя, но с другой стороны код может вызываться любое количество раз 

In [ ]:
i = 0
while i < 10:
    print(i)
    i += 1

In [ ]:
# Если нужно пробежаться определённое количество раз (или пробежаться по элементам коллекции, что будет позже),
#   то удобнее пользоваться конструкцией for

for i in range(10):
    print(i)

In [10]:
# Пусть есть банковский вклад, а проценты начисляются без капитализацииz
#  Распечатаем какая будет сумма после каждого прощедшего года.
#  Предполжим, что мы не знаем математику и соответствующую формулу, зато знаем циклы :)

START_CAPITAL = 100
n_years = 3
interest = 0.1

CAPITAL = START_CAPITAL
for i in range(n_years):
  CAPITAL += START_CAPITAL * interest
  print(f"После {i} года капитал составляет {CAPITAL} тугриков.")

После 0 года капитал составляет 110.0 тугриков.
После 1 года капитал составляет 120.0 тугриков.
После 2 года капитал составляет 130.0 тугриков.


In [ ]:
# Задача: напишите, а затем вызовите несколько раз функцию, которая расчитывает объем вклада по истечению определенного количества лет.


## Составные типы данных

Выше мы говорили о переменных, которые могут хранить числа и строки, но что еси нужно хранить сразу несколько значений? Для этого предназначены составные типы данных: списки, кортежи, словари, множества.

### Списки (lists)

Списки создаются с помощью квадратных скобок.

In [ ]:
five_element_list = [1,2,3,4,5]
empty_list = []

In [ ]:
# Функция len позволяет получить длину сложных объектов
print(len(five_element_list))
print(len(empty_list))

In [ ]:
# Получить доступ к элементу списка можно по индексу с помощью квадратных скобок
#  Самый первый элемент имеет индес 0
print(five_element_list[0])
# Для нумерации списка с конца можно использовать отрицательные индексы
#  Последний с конце имеет индекс -1, а следующий с конца -- -2 и т.д.
print(five_element_list[-2])

# Либо сразу несколько элементов
print(five_element_list[1:3])

In [ ]:
# В списке могут лежать элементы разных типов одновременно
hetero_list = [1, "текст 2", 3]
print(hetero_list)

In [ ]:
# По индексу можно менять значения
five_element_list[3] = 33 # Нумерация с нуля!
print(five_element_list)

In [ ]:
# Можно в конец списка добавить новый элемент
five_element_list.append(33)
five_element_list.append(1)
print(five_element_list)

In [ ]:
# Можно удалять первый найденный элемент по значению
five_element_list.remove(33)
print(five_element_list)

In [ ]:
# Можно удалять первый элемент по значению
five_element_list.remove(33)
print(five_element_list)

In [ ]:
# Можно удалять первый элемент по значению
five_element_list.remove(33)
print(five_element_list)

### Словари

Словарь -- структура данных, в которых элементы доступны по ключю (уникальному имени элемента), при этом ключ у каждого элемента должен быть уникальным. Для списка, подобным ключем являлся индекс элемента.

In [ ]:
# Создаем словарь

three_elem_dict = {1:'value1', 5:'value2', 'key3':'value3'}
print(three_elem_dict)

In [ ]:
# Получаем элемент по ключу
print(three_elem_dict[1])

In [ ]:
# Если элемента нет, то ошибка
print(three_elem_dict['1'])

In [ ]:
# Вернём значение по умолчанию, если нет
print(three_elem_dict.get('1',None))
print(three_elem_dict.get(1,None))

In [ ]:
# Добавляем новый элемент с новым ключем
three_elem_dict['key4']='value4'
print(three_elem_dict)

In [ ]:
# Удаляем элемент
del(three_elem_dict['key3'])
print(three_elem_dict)

In [ ]:
# Сохраняем то что удаляем отдельно
removed = three_elem_dict.pop('key4')
print(removed)
print(three_elem_dict)

In [ ]:
# Список ключей
print(three_elem_dict.keys())

### ВАЖНО (!): присвоение элементов со сложными объектами

При присвоении одной переменной значения другой переменной, обе переменные указывают на одну(!) область памяти, а значит при редактировании одной переменной редактируется и другая.

In [ ]:
a = {1:'1',2:'2'}
b = a
print(a, b)
b[1]='11'
print(a, b)

In [ ]:
a = [[1,2],[2,3]]
for el in a:
    el[0] = 100
print(a)

In [ ]:
# Если нужна копия, то можно использовать один из следующих вариантов
a = {1:'1',2:'2'}
b = dict(a)
c = a.copy()
print(a,b,c)
a[1]='11'
print(a,b,c)

## Циклы: продолжение

Возможно проводить циклы по коллекциям.

In [ ]:
l = [1,2,3,4,"five",{1:"1",2:"2"}]
for e in l:
    print(e)

1
2
3
4
five
{1: '1', 2: '2'}


In [ ]:
d = {1:"1",2:"2","three":[1,2]}
for k in d:
    print(k)

1
2
three


In [ ]:
# А также с ключем
for k,v in d.items():
    print("Ключ:",k, "Значение:", v)

Ключ: 1 Значение: 1
Ключ: 2 Значение: 2
Ключ: three Значение: [1, 2]


### Задание:
1. Создайте произвольный список из 10 элементов с повторяющимися элементами
1. Создайте словарь, в котором ключами будут элементы из созданного списка, а значениями -- количество повторов каждого элемента

Часто циклы занимают много вычислительных ресурсов, поэтому для повышения эффективности вычислений и повышения читаемости кода бывает нужно либо остановить весь цикл или пропустить текущую итерацию.

In [ ]:
# Пусть функция проверки високосности года является "меделенной"
# Нужно проверить, есть ли хотя бы один високосный год в списке

years = [1999, 2001, 2004, 2000, 2100]
has_leap_year = False
for y in years:
    print(y)
    if is_leap_year(y):
        # Точно есть хотя бы один
        has_leap_year = True
        break
print(has_leap_year)

1999
2001
2004
True


In [ ]:
# Однако, возможно, в некоторых случаях вы можете быстро проверить, что год не високосный, 
#   например, в случае если год нечётный

years = [1999, 2001, 2004, 2000, 2100]
has_leap_year = False
for y in years:
    print("Начало:", y)
    if y % 2 == 1:
        # Точно не високосный
        continue
    print("Проверка:", y)
    if is_leap_year(y):
        # Точно есть хотя бы один
        has_leap_year = True
        break
print(has_leap_year)

Начало: 1999
Начало: 2001
Начало: 2004
Проверка: 2004
True


### Задание: циклы

Необходимо написать функцию, которая проверяет, что в списке целых чисел есть простое число.

Помните, что функция проверки на простое число -- достаточно медленная. Также помните, что чётное число может быть простым, только если это 2.

In [ ]:
# ...

# Классы

Классы это структура данных со своими методами. Можно создать несколько независимых объектов одного класса. У каждогой объект будет свое состояние

In [47]:
class Worker:
  def __init__(self, name):
    self.name = name
    print(f"Работник {self.name} родился!")

  def __call__(self):
    print(f"Работник {self.name}: работу работаю на работе...")

class Boss:
  def __init__(self, name):
    self.name = name
    print(f"Бос {self.name} родился!")

  def __call__(self):
    print(f"Бос {self.name}: Работать, работать, работать!")

  def give_some_lyuley(self, worker):
    if not isinstance(worker, Worker):
      print(f"Бос {self.name}: люлей выдаю только работникам!")
    else:
      print(f"Бос {self.name} выдал люлей работнику {worker.name}.")

class BigBoss(Boss):
  def __init__(self, name):
    super(BigBoss, self).__init__(name)
    self.name = name
    print(f"БольшойБос {self.name} родился!")

  def give_some_lyuley(self, worker):
    if isinstance(worker, Boss):
      print(f"BigБос {self.name} выдал люлей босу {worker.name}.")
    elif isinstance(worker, Worker):
      print(f"BigБос {self.name} выдал люлей работнику {worker.name}.")
    elif isinstance(worker, BigBoss):
      print(f"BigБос {self.name}: больших начальников не ругаю!")
    else:
      print(f"BigБос {self.name}: а это что за тюлень!?")
  


In [48]:
w1 = Worker("w1")
w2 = Worker("w2")

Работник w1 родился!
Работник w2 родился!


In [49]:
b1 = Boss("B1")
b2 = Boss("B2")

Бос B1 родился!
Бос B2 родился!


In [50]:
bb1 = BigBoss("BB1")
bb2 = BigBoss("BB2")

Бос BB1 родился!
БольшойБос BB1 родился!
Бос BB2 родился!
БольшойБос BB2 родился!


In [51]:
b1()
b1.give_some_lyuley(w1)

Бос B1: Работать, работать, работать!
Бос B1 выдал люлей работнику w1.


In [52]:
bb1()
bb1.give_some_lyuley(b1)

Бос BB1: Работать, работать, работать!
BigБос BB1 выдал люлей босу B1.


Рассмотрим класс списка

In [ ]:
class MyList:
    def __init__(self,s):
        self.__s=list(s)
    def append(self, new_elem):
        self.__s.append(new_elem)
    def get(self, i):
        return self.__s[i]
    def print(self):
        print(self.__s)

In [ ]:
a = MyList([1,2,3])
a.print()
print(a.append(5))
print(a.get(2))
a.print()

[1, 2, 3]
None
3
[1, 2, 3, 5]


### Задание: классы

Напишите небольшой класс с почти теми же методами, что и класс выше, который имитирует множество.
В качестве **__s** используйте также список, но при этом в **append** необходимо проверить, что такого элемента в множестве нет, а вместо функции **get** напишите функцию **has**, проверяющую, что элемент есть во множестве. 

In [ ]:
# подсказка
a=[1,2,3]
b = 2
print(b in a)

True


In [ ]:
# ...

## Модули и пространства имен

Модули необходимы для передачи кода между людьми. В дальнейшем мы будем пользоваться такими модулями как **pandas**, **numpy**, **matplotlib**, **sklearn**, которые реализуют за нас оснонвые операции по работе и анализу данных.

Для демонстрации создадим файлmy_list.py либо рядом с блокнотом, либо в системной паке. 

А также поместим его в подпапку tmp.

Его содержимое представлено ниже:

In [ ]:
class MyList:
    def __init__(self,s):
        self.__s=list(s)
    def append(self, new_elem):
        self.__s.append(new_elem)
    def get(self, i):
        return self.__s[i]
    def print(self):
        print("MyList:", self.__s)
        
class MyList2:
    def __init__(self,s):
        self.__s=list(s)
    def append(self, new_elem):
        self.__s.append(new_elem)
    def get(self, i):
        return self.__s[i]
    def print(self):
        print("MyList2:", self.__s)

Мы можем воспользоваться классами определёнными в этом файле, если импортируем его. Будем его рассматривать как отдельный модуль.

In [55]:
import my_list

l1 = my_list.MyList([1,2])
l2 = my_list.MyList2([1,2])
l1.print()
l2.print()

MyList: [1, 2]
MyList2: [1, 2]


In [56]:
# Можно импортировать только один класс
from my_list import MyList

l1 = MyList([1,2])
l1.print()
l2 = MyList2([1,2])
l2.print()

MyList: [1, 2]


NameError: ignored

In [57]:
# А также можно переименовать модуль или импортируемый объект
import my_list as m_l

l1 = m_l.MyList([1,2])
l2 = m_l.MyList2([1,2])
l1.print()
l2.print()

MyList: [1, 2]
MyList2: [1, 2]


In [58]:
# Можно импортировать только один класс
from my_list import MyList as ML

l1 = ML([1,2])
l1.print()

MyList: [1, 2]


In [59]:
import tmp

In [66]:
from tmp import my_list as tmp_mylist
tmp_mylist.MyList([1,2]).print()

MyList: [1, 2]


### Задание: модули и пространства имен

Скопируйте в отдельный файл my_set.py описанный вами класс. Загрузите его в этот блокнот разными, описанными выше способами и создайте соответсвующий объект.

In [ ]:
# ...

# Продвинутые темы

## Функция как переменная

Функция может рассматриваться как переменная.

In [ ]:
# Предположим, мы хотим применить функцию сразу ко всем элементами списка

def x2(x):
    return x**2

x22 = x2

def apply_f(l, f):
    res = []
    for e in l:
        res.append(f(e))
    return res

print(apply_f([1,2,3],x2))
print(apply_f([1,2,3],x22))

[1, 4, 9]
[1, 4, 9]


In [ ]:
# Проверка на чётность

def is_even(x):
    return x % 2 == 1
print(apply_f([1,2,3],is_even))

[True, False, True]


### Задание: функция как переменная

Распечатайте результат проверки каждого числа в списке на простоту (простые числа)

In [ ]:
# ...

## lambda функции

In [ ]:
# Функцию x2 можно переписать через lambda, т.е. неименованную функцию

x2 = lambda x: x**2
print(apply_f([1,2,3],x2))

# или сразу напрямую
print(apply_f([1,2,3], lambda x: x**2))

[1, 4, 9]
[1, 4, 9]


### Задание: lambda функцию

Через lambda функцию переведите список из чисел от 0 до 1 в проценты в виде строки от "0%" до "100%".

In [ ]:
# Пусть есть число
x = 0.2
# Тогда в процентах это будет
print(str(x*100)+"%")

# Задайте произвольный список из чисел от 0 до 1 
#   и замените его на процентную запись с помощью lambda функции
# ...

20.0%


## list comprehension (создание списков)



In [ ]:
x_list = list(range(1,25))
print(x_list)

# Вместо функции apply_f, используем конструкции языка
res = [x2(x) for x in x_list]
print(res)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24]
[1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400, 441, 484, 529, 576]


In [ ]:
# При этом можно отбирать только некоторые элементы, например, кратные тройке
res = [x for x in x_list if x % 3 == 0]
print(res)
# Либо ещё и применить какую-то функцию
res = [x**2 for x in x_list if x % 3 == 0]
print(res)

[3, 6, 9, 12, 15, 18, 21, 24]
[9, 36, 81, 144, 225, 324, 441, 576]


Также с списки можно генерить на основании других "итерируемых" объектов, например, файлов.

Пусть если файл [log.txt](https://yadi.sk/d/vTnh0MtalzSLBg) со следующим содержимым (первая колонка -- IP, вторая количество байт):

> 127.0.0.1 120

> 10.1.1.1 210

> 127.0.0.1 80

> 10.1.1.1 215

> 10.1.1.1 200

> 10.1.1.2 210

In [ ]:
# Читаем файл и сразу разделяем его на колонки
raw = [line.split(" ")  for line in open("log.txt")]
print(raw)

res = {}
for k,v in raw:
    if k in res:
        res[k] += int(v)
    else:
        res[k] = int(v)
print(res)

[['127.0.0.1', '120\n'], ['10.1.1.1', '210\n'], ['127.0.0.1', '80\n'], ['10.1.1.1', '215\n'], ['10.1.1.1', '200\n'], ['10.1.1.2', '210\n']]
{'127.0.0.1': 200, '10.1.1.1': 625, '10.1.1.2': 210}


### Задание: list comprehension

Из заданного списка целых чисел выбирите все  простые числа.

In [ ]:
# ...

## Генераторы

По сути, внтреняя часть list comprehension, т.е. то, что находится между квадратными скобками, является генератором, а квадратные скобки "пробегаются" по этому генератору и заполняют список. Поэтому сам генератор занимает очень мало памяти в компьютере и является просто правилом как создать элементы.

Так, например, если файл-лог очень большой, мы не хотим его сразу подгружать в память и можем просто создать генератор.

In [ ]:
raw_gen = (line.split(" ")  for line in open("log.txt"))
# Память не занята, мы просто знаем как получать строки.
#  и только тут мы будем эти строки читать, то не сразу все, а одна за одной.
res = {}
for k,v in raw_gen:
    if k in res:
        res[k] += int(v)
    else:
        res[k] = int(v)
print(res)

{'127.0.0.1': 200, '10.1.1.1': 625, '10.1.1.2': 210}


In [ ]:
# Также можно писать свои функции-генераторы

def all_even_numbers():
    i = 2;
    while(True):
        yield i;
        i = i + 2
        
for even_num in all_even_numbers():
    if even_num > (20):
        break
    print(even_num)

2
4
6
8
10
12
14
16
18
20


### Сортировка

In [ ]:
sorted([1,3,2])

[1, 2, 3]

# Задачи

Далее даны примеры задач, которые могут помочь заточить навык работы на Python. Рекомендуется оформлять задачу разными способами. В частности, когда возможно, стоит использовать функции, генераторы, списки, lambda -- функции и так далее.

## Задача 1

Для произвольного списка чисел **l**  
и числа **a**,
вывести все элементы **l**
меньшие **a**.

## Задача 2

По двум спискам **l**, **m** создать третий список состоящий из общих элементов для **l**, **m**

## Задача 3

Для произвольного словаря **d** и числа **n** выведите **n** ключей словаря **d**, соответствующих максимальным по значению элементам словаря

## Задача 4

Вывести первые **n** строк треугольника Паскаля. В треугольнике в вершине и по сторонам стоят единицы, в каждой следующей строке на один элемент больше, центральные элементы равны сумме элементов над ним.

Например первые 5 строк:
1. 1
1. 1 1
1. 1 2 1
1. 1 3 3 1
1. 1 4 6 4 1

## Задача 5

Напишите проверку на то, является ли строка палиндромом. Палиндром — это слово или фраза, которые одинаково читаются слева направо и справа налево.

## Задача 6

Напишите проверку на то, является ли число палиндромом.

## Задача 7

Отобразите число секунд **n** в виде дни:часы:минуты:секунды

> 3700 --> 0 дней : 1 час : 1 минута : 40 секунд

## Задача 8

При заданном целом числе n посчитайте n + nn + nnn. Например, n=12, тогда 12 + 1212 + 121212 = 122436

## Задача 9

Сложите цифры целого числа n.

## Задача 10

Нужно проверить, все ли числа в последовательности уникальны.

![Заключительная картинка](end-image.png)